src/java/org/apache/cassandra/db/Mutation.java

```diff
@@ -150,6 +150,11 @@ public ImmutableCollection<PartitionUpdate> getPartitionUpdates()
         return modifications.values();
     }
 
+    public long getApproxCreatedAtNanos()
+    {
+        return approxCreatedAtNanos;
```

Comment:
`ModificationStatement#executeWithoutCondition` creates the write/delete timestamps before the mutations themselves are created. This means that hints could still be delivered late if enough time passes after creating the timestamps. 

This should err on the side of discarding hints early, so we should be recording a timestamp before actually creating mutation timestamps.

Reply:
Yeah definitely agree that there is still this window(albeit likely much shorter) between mutation timestamp creation and the approximate time generated inside the mutation constructor.
At first glance it's quite a bit more fiddly to get some sort of timestamp pre mutation constructor timestamp and propagate it everywhere, but I can have a go at taking RequestTime.startedAtNanos or RequestTime.enqueuedAtNanos and plumbing it through to the mutation constructor everywhere. 

We'll also be adding a long in the Mutation object to solve this issue, so at fairly tiny but non-negligible cost.
Or I suppose slightly alternatively we could try and pull approxCreatedAtNanos out to just before the write/delete timestamps are generated, though this has some downstream effects due to it's other use, which need to be reasoned carefully about.

Reply:
I probably wouldn't add it to Mutation. For `ModificationStatement#executeWithoutCondition` at least, you can grab nanoTime before the call to getMutations, then just pass it into the StorageProxy call as an argument. From there I think you need to plumb it to AbstractWriteResponseHandler and the call to `hintMutations`